ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -0,0 +1,364 @@
+     *
+     * Efficient way is to just pass the src input-stream to azure-sdk as shown below.
+     * messageSievingInputStream has a list of input-streams already, where each stream is a blob.
+     *
+     *  If CloudBlobStore.put(messageWriteSet) can directly call AzureCloudDestinationSync.uploadBlob(messageWriteSet),
+     *  then we can do this here:
+     *
+     *     MessageSievingInputStream messageSievingInputStream =
+     *         (MessageSievingInputStream) ((MessageFormatWriteSet) messageSetToWrite).getStreamToWrite();
+     *     List<InputStream> messageStreamList = messageSievingInputStream.getValidMessageStreamList();
+     *     // Each input stream is a blob
+     *     ListIterator<InputStream> messageStreamListIter = messageStreamList.listIterator();
+     *     // Pass the stream to azure-sdk, no need of a memcpy
+     *     BlobParallelUploadOptions blobParallelUploadOptions = new BlobParallelUploadOptions(messageStreamListIter.next());
+     *
+     */
+    Timer.Context storageTimer = null;
+    // setNameSchemeVersion is an ugly remnant of legacy code. We have to set it explicitly.
+    cloudBlobMetadata.setNameSchemeVersion(azureCloudConfig.azureNameSchemeVersion);
+    // Azure cloud container names must be 3 - 63 char
+    AzureBlobLayoutStrategy.BlobLayout blobLayout = azureBlobLayoutStrategy.getDataBlobLayout(cloudBlobMetadata);
+    String blobIdStr = blobLayout.blobFilePath;
+    try {
+      // Prepare to upload blob to Azure blob storage
+      // There is no parallelism, but we still need to create and pass this object to SDK.
+      BlobParallelUploadOptions blobParallelUploadOptions =
+          new BlobParallelUploadOptions(blobInputStream);
+      // To avoid overwriting, pass "*" to setIfNoneMatch(String ifNoneMatch)
+      // https://learn.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable
+      blobParallelUploadOptions.setRequestConditions(new BlobRequestConditions().setIfNoneMatch("*"));
+      // This is ambry metadata, uninterpreted by azure
+      blobParallelUploadOptions.setMetadata(cloudBlobMetadata.toMap());
+      // Without content-type, get-blob floods log with warnings
+      blobParallelUploadOptions.setHeaders(new BlobHttpHeaders().setContentType("application/octet-stream"));
+
+      BlobContainerClient blobContainerClient = createOrGetBlobStore(blobLayout.containerName);
+      ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
+      storageTimer = azureMetrics.blobUploadTime.time();
+      Response<BlockBlobItem> blockBlobItemResponse =
+          blobContainerClient.getBlobClient(blobIdStr)
+              .uploadWithResponse(blobParallelUploadOptions, Duration.ofMillis(cloudConfig.cloudRequestTimeout),
+                  Context.NONE);
+      ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
```

Comment:
what about retries? are we going to implement that later?

Reply:
Retries are handled by azure sdk @justinlin-linkedin

Reply:
Retries are handled by azure sdk @justinlin-linkedin.
`createBlobStorageSyncClient`